File: Week_06/java/src/besttimetobuyandsellstock/_121_BestTimeToBuyAndSellStock.java
package besttimetobuyandsellstock;

/**
 * 第一遍：2020/07/30周四 ✅
 * 第二遍：2020/07/30周五 ✅
 * 第三遍：2020/07/29周三
 * 第四遍：2020/07/05周日
 */
class _121_BestTimeToBuyAndSellStock {
    public int maxProfit(int[] prices) {
        //f(i)表示第i天卖出时可能获得的最大利润
        //卖出价固定，买入价越低越好，在扫描到第i天时我们记录前i-1天中的最低价格
        //i in [1, n - 1] f(0) = 0
        if (prices == null || prices.length < 2) {
            return 0;
        }
        int maxDiff = 0;
        int n = prices.length;
        int minPrice = prices[0];
        for (int i = 1; i < n; i++) {
            int currDiff = prices[i] - minPrice;
            if (currDiff > maxDiff) {
                maxDiff = currDiff;
            }
            if (prices[i] < minPrice) {
                minPrice = prices[i];
            }
        }
        return maxDiff;
    }

    class SolutionDp {
        public int maxProfit(int[] prices) {
            //f(i)表示第i天可能获得的最大利润
            //f(i) = f(i - 1) + prices[i] - prices[i - 1]
            //i in [1, n - 1] f(0) = 0
            if (prices == null || prices.length < 2) {
                return 0;
            }
            int dp = 0;
            int n = prices.length;
            int maxProfit = 0;
            for (int i = 1; i < n; i++) {
                int currDiff = prices[i] - prices[i - 1];
                dp = Math.max(0, dp + currDiff);

                if (dp > maxProfit) {
                    maxProfit = dp;
                }
            }
            return maxProfit;
        }
    }
}

File: Week_06/kotlin/src/besttimetobuyandsellstock/_121_BestTimeToBuyAndSellStock.kt
package besttimetobuyandsellstock

fun maxProfitDp(prices: IntArray): Int {
    //f(i)(0)表示第i天未持有股票的最大利润
    //f(i)(1)表示第i天持有股票的最大利润
    //f(i)(0) = Math.max(f(i - 1)(0), f(i - 1)(1) + prices[i])
    //f(i)(1) = Math.max(f(i - 1)(1),  - prices[i])
    //f(0)(0) = 0, f(0)(1) = Integer.MIN_VALUE
    if (prices.size < 2) {
        return 0
    }
    val n = prices.size
    val dp = Array(n) { IntArray(2) }
    dp[0][0] = 0
    dp[0][1] = -prices[0]
    for (i in 1 until n) {
        dp[i][0] = dp[i - 1][0].coerceAtLeast(dp[i - 1][1] + prices[i])
        dp[i][1] = dp[i - 1][1].coerceAtLeast(-prices[i])
    }
    return dp[n - 1][0].coerceAtLeast(dp[n - 1][1])
}

fun maxProfit(prices: IntArray): Int {
    //f(i)表示第i天卖出股票时可以获得的最大利润
    //记录i天之前股票的最低价格
    //f(0) = 0 f(1) = prices[1] - minPrices
    if (prices.size < 2) {
        return 0
    }
    var maxProfit = 0
    var minPrice = prices[0]
    for (i in 1 until prices.size) {
        maxProfit = maxProfit.coerceAtLeast(prices[i] - minPrice)
        minPrice = minPrice.coerceAtMost(prices[i])
    }
    return maxProfit
}

fun maxProfit2(prices: IntArray): Int {
    //f(i)表示到第i天可以获得的最大利润(大于0的),不是从第0天开始，而是从j天开始，类似子数组的最大值
    //f(i) = max(0,f(i - 1) + prices[i] - prices[i - 1])
    //f(0) = 0
    if (prices.size < 2) {
        return 0
    }
    var maxProfit = 0
    var profit = 0
    for (i in 1 until prices.size) {
        profit = 0.coerceAtLeast(profit + prices[i] - prices[i - 1])
        maxProfit = maxProfit.coerceAtLeast(profit)
    }
    return maxProfit
}